engine/e_bluefield.c

```diff
@@ -829,6 +924,728 @@ static int engine_pka_dh_finish(DH *dh)
+                          adata->pubk_bytes);
+}
+
+static int engine_pka_curve448_bits(const EVP_PKEY *pkey)
+{
+    return PKA_CURVE448_BITS;
+}
+
+static int engine_pka_curve448_security_bits(const EVP_PKEY *pkey)
+{
+    return PKA_CURVE448_SECURITY_BITS;
+}
+
+static int engine_pka_curve25519_bits(const EVP_PKEY *pkey)
+{
+    return PKA_CURVE25519_BITS;
+}
+
+static int engine_pka_curve25519_security_bits(const EVP_PKEY *pkey)
+{
+    return PKA_CURVE25519_SECURITY_BITS;
+}
+
+static int engine_pka_cmp_parameters(const EVP_PKEY *a, const EVP_PKEY *b)
+{
+    return 1;
+}
+
+static int
+engine_pka_gen_ctrl(int nid, EVP_PKEY *pkey, int op, long arg1, void *arg2)
+{
+    ENGINE_PKA_KEYPAIR *kp = NULL;
+    const unsigned char *p = NULL;
+    const struct engine_pka_nid_data_st *nid_data = engine_pka_get_nid_data(nid);
+    int pklen = 0;
+
+    switch (op)
+    {
+
+#ifndef OPENSSL_V102_COMPAT
+    case ASN1_PKEY_CTRL_SET1_TLS_ENCPT:
+        p = arg2;
+        pklen = arg1;
+
+        if (p == NULL || pklen != nid_data->pubk_bytes )
+        {
+            printf("ERROR: Wrong length\n");
+            return 0;
+        }
+
+        kp = engine_pka_keypair_new(nid, PKA_NO_PRIV_KEY, pklen);
+        if (engine_pka_keypair_invalid(kp, nid, 0))
+        {
+            printf("ERROR: Invalid keypair\n");
+            return 0;
+        }
+
+        memcpy(kp->public_key.buf_ptr, p, pklen);
+
+        EVP_PKEY_assign(pkey, nid, kp);
+        return 1;
+
+
+    case ASN1_PKEY_CTRL_GET1_TLS_ENCPT:
+        kp = EVP_PKEY_get0(pkey);
+        if (!engine_pka_keypair_invalid(kp, nid, 0) && nid == kp->nid)
+        {
+            unsigned char **ppt = arg2;
+            *ppt = OPENSSL_memdup(kp->public_key.buf_ptr, nid_data->pubk_bytes);
+            if (*ppt != NULL)
+                return nid_data->pubk_bytes;
+        }
+        return 0;
+#endif
+    case ASN1_PKEY_CTRL_DEFAULT_MD_NID:
+        *(int *)arg2 = NID_undef;
+        return 2;
+
+    default:
+        return -2;
+
+    }
+}
+
+
+static int
+engine_pka_gen_priv_encode(int nid, PKCS8_PRIV_KEY_INFO *p8,
+                           const EVP_PKEY *pkey)
+{
+    const ENGINE_PKA_KEYPAIR *kp = EVP_PKEY_get0(pkey);
+    ASN1_OCTET_STRING oct;
+    unsigned char *penc = NULL;
+    int penclen;
+    const struct engine_pka_nid_data_st *nid_data = engine_pka_get_nid_data(nid);
+    char *tmp_buf = NULL;
+    int ret = 0;
+
+    if (nid_data == NULL)
+    {
+        printf("ERROR: Missing NID data\n");
+        return 0;
+    }
+
+    if (engine_pka_keypair_invalid(kp, nid, 1) || kp->nid != nid)
+    {
+        printf("ERROR: Invalid private key\n");
+        return 0;
+    }
+
+    tmp_buf = OPENSSL_secure_malloc(nid_data->privk_bytes);
+    if (NULL == tmp_buf)
+    {
+        printf("ERROR: OpenSSL malloc failed\n");
+        return 0;
+    }
+
+    oct.data = memcpy(tmp_buf, kp->private_key.buf_ptr, nid_data->privk_bytes);
```

Comment:
Q: we are storing the return code of memcpy into oct.data; is it expected?

Reply:
Yes, `memcpy()` returns the pointer to destination_ptr, which in this case is `tmp_buf`. `oct` is further used below to encode private key into ASN1 format.

Reply:
Ack